File: python/app/catalog.py
from datetime import datetime, timezone
from time import perf_counter
import math
import httpx


def normalize_model(raw: dict) -> dict:
    architecture = raw.get("architecture", {})
    inputs, outputs = (
        architecture.get("input_modalities", []),
        architecture.get("output_modalities", []),
    )
    exclusion = (
        "Not a text-in, text-out model."
        if "text" not in inputs or "text" not in outputs
        else "A routing alias; this demo selects the underlying model itself."
        if raw["id"].startswith("openrouter/")
        else None
    )

    def price(key):
        value = raw.get("pricing", {}).get(key)
        return (
            float(value) * 1_000_000
            if value is not None and math.isfinite(float(value)) and float(value) >= 0
            else None
        )

    return dict(
        id=raw["id"],
        name=raw["name"],
        description=raw.get("description", "")[:500],
        contextLength=raw.get("context_length", 0),
        inputPrice=price("prompt"),
        outputPrice=price("completion"),
        inputModalities=inputs,
        outputModalities=outputs,
        eligible=not exclusion,
        exclusion=exclusion,
        reasoning="reasoning" in raw.get("supported_parameters", []),
    )


async def fetch_catalog() -> dict:
    started = perf_counter()
    async with httpx.AsyncClient(timeout=20) as client:
        response = await client.get("https://openrouter.ai/api/v1/models")
        response.raise_for_status()
    models = response.json()["data"]
    if not models:
        raise ValueError("OpenRouter returned an empty catalog.")
    return dict(
        stage="catalog",
        models=[normalize_model(m) for m in models],
        fetchedAt=datetime.now(timezone.utc).isoformat(),
        durationMs=round((perf_counter() - started) * 1000),
    )

File: typescript/src/catalog.ts
import type { Catalog, Model } from "../../shared/types";

export function normalizeModel(raw: any): Model {
  const input = raw.architecture?.input_modalities ?? [];
  const output = raw.architecture?.output_modalities ?? [];
  const exclusion =
    !input.includes("text") || !output.includes("text")
      ? "Not a text-in, text-out model."
      : raw.id.startsWith("openrouter/")
        ? "A routing alias; this demo selects the underlying model itself."
        : null;
  const price = (value: unknown) =>
    value !== null &&
    value !== undefined &&
    Number.isFinite(Number(value)) &&
    Number(value) >= 0
      ? Number(value) * 1_000_000
      : null;
  return {
    id: raw.id,
    name: raw.name,
    description: (raw.description ?? "").slice(0, 500),
    contextLength: raw.context_length ?? 0,
    inputPrice: price(raw.pricing?.prompt),
    outputPrice: price(raw.pricing?.completion),
    inputModalities: input,
    outputModalities: output,
    eligible: !exclusion,
    exclusion,
    reasoning: (raw.supported_parameters ?? []).includes("reasoning"),
  };
}

export async function fetchCatalog(): Promise<Catalog> {
  const started = performance.now();
  const response = await fetch("https://openrouter.ai/api/v1/models", {
    signal: AbortSignal.timeout(20_000),
    cache: "no-store",
  });
  if (!response.ok)
    throw new Error(
      `OpenRouter catalog unavailable (HTTP ${response.status}).`,
    );
  const data = await response.json();
  if (!Array.isArray(data.data) || !data.data.length)
    throw new Error("OpenRouter returned an empty catalog.");
  return {
    stage: "catalog",
    models: data.data.map(normalizeModel),
    fetchedAt: new Date().toISOString(),
    durationMs: Math.round(performance.now() - started),
  };
}
